File: cdn-system/api/services/node_config_service.go
package services

import (
	"cdn-api/db"
	"cdn-api/models"
	"errors"
	"time"

	"gorm.io/gorm"
)

func UpsertNodeConfigItem(nodeID int64, name string, value string) error {
	if nodeID == 0 || name == "" {
		return nil
	}
	var item models.ConfigItem
	query := db.DB.Where("type = ? AND scope_name = ? AND scope_id = ? AND name = ?", "node_config", "node", nodeID, name)
	if err := query.First(&item).Error; err == nil {
		updates := map[string]interface{}{
			"value":     value,
			"enable":    true,
			"update_at": time.Now(),
		}
		return query.Model(&models.ConfigItem{}).Updates(updates).Error
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return err
	}

	item = models.ConfigItem{
		Name:      name,
		Value:     value,
		Type:      "node_config",
		ScopeName: "node",
		ScopeID:   nodeID,
		Enable:    true,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return db.DB.Create(&item).Error
}

func GetNodeConfigMap(name string) (map[int64]string, error) {
	result := map[int64]string{}
	if name == "" {
		return result, nil
	}
	var items []models.ConfigItem
	if err := db.DB.Where("type = ? AND scope_name = ? AND name = ?", "node_config", "node", name).Find(&items).Error; err != nil {
		return nil, err
	}
	for _, item := range items {
		result[item.ScopeID] = item.Value
	}
	return result, nil
}

func GetNodeConfigValue(nodeID int64, name string) (string, error) {
	if nodeID == 0 || name == "" {
		return "", nil
	}
	var item models.ConfigItem
	if err := db.DB.Where("type = ? AND scope_name = ? AND scope_id = ? AND name = ?", "node_config", "node", nodeID, name).First(&item).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return "", nil
		}
		return "", err
	}
	return item.Value, nil
}

File: cnn.net/src/Cnn.Api/Services/Common/NodeConfigService.cs
using Cnn.Domain.Entities;
using SqlSugar;
using Task = System.Threading.Tasks.Task;

namespace Cnn.Api.Services.Common;

public interface INodeConfigService
{
    Task UpsertAsync(long nodeId, string name, string value, CancellationToken cancellationToken);
    Task<Dictionary<long, string>> GetMapAsync(string name, CancellationToken cancellationToken);
    Task<string?> GetValueAsync(long nodeId, string name, CancellationToken cancellationToken);
}

public sealed class NodeConfigService : INodeConfigService
{
    private const string ConfigType = "node_config";
    private const string ScopeName = "node";

    private readonly ISqlSugarClient _db;

    public NodeConfigService(ISqlSugarClient db)
    {
        _db = db;
    }

    public async Task UpsertAsync(long nodeId, string name, string value, CancellationToken cancellationToken)
    {
        if (nodeId <= 0 || string.IsNullOrWhiteSpace(name))
        {
            return;
        }

        var item = await _db.Queryable<Config>()
            .Where(c => c.Type == ConfigType && c.ScopeName == ScopeName && c.ScopeId == nodeId && c.Name == name)
            .FirstAsync();

        var now = DateTime.Now;
        if (item != null)
        {
            await _db.Updateable<Config>()
                .SetColumns(c => new Config
                {
                    Value = value,
                    Enable = true,
                    UpdateAt = now
                })
                .Where(c => c.Type == ConfigType && c.ScopeName == ScopeName && c.ScopeId == nodeId && c.Name == name)
                .ExecuteCommandAsync();
            return;
        }

        var defaultTaskId = _db.CurrentConnectionConfig?.DbType == DbType.Sqlite
            ? 0L
            : (long?)null;

        await _db.Insertable(new Config
        {
            Name = name,
            Value = value,
            Type = ConfigType,
            ScopeName = ScopeName,
            ScopeId = (int)nodeId,
            Enable = true,
            // MySQL schema enforces FK on task_id, while SQLite test schema expects a non-null default.
            TaskId = defaultTaskId,
            CreateAt = now,
            UpdateAt = now
        }).ExecuteCommandAsync();
    }

    public async Task<Dictionary<long, string>> GetMapAsync(string name, CancellationToken cancellationToken)
    {
        var result = new Dictionary<long, string>();
        if (string.IsNullOrWhiteSpace(name))
        {
            return result;
        }

        var items = await _db.Queryable<Config>()
            .Where(c => c.Type == ConfigType && c.ScopeName == ScopeName && c.Name == name)
            .ToListAsync();

        foreach (var item in items)
        {
            if (item.ScopeId.HasValue)
            {
                result[item.ScopeId.Value] = item.Value ?? string.Empty;
            }
        }

        return result;
    }

    public async Task<string?> GetValueAsync(long nodeId, string name, CancellationToken cancellationToken)
    {
        if (nodeId <= 0 || string.IsNullOrWhiteSpace(name))
        {
            return null;
        }

        var item = await _db.Queryable<Config>()
            .Where(c => c.Type == ConfigType && c.ScopeName == ScopeName && c.ScopeId == nodeId && c.Name == name)
            .FirstAsync();

        return item?.Value;
    }
}
